play-json/shared/src/main/scala/play/api/libs/json/JsValue.scala

```diff
@@ -31,15 +31,24 @@ object JsValue {
 /**
  * Represents a Json null value.
  */
-case object JsNull extends JsValue
+object JsNull extends JsValue {
+  override def equals(that: Any): Boolean = that match {
+    case _: JsNull.type => true
+    case other: JsValue => other.equals(this)
+    case _ => false
+  }
+}
 
 /**
  * Represents a Json boolean value.
  */
 sealed abstract class JsBoolean(
   val value: Boolean) extends JsValue with Product with Serializable {
 
-  def canEqual(that: Any): Boolean = that.isInstanceOf[JsBoolean]
+  def canEqual(that: Any): Boolean = that match {
+    case _: JsBoolean => true
+    case _ => false
+  }
```

Comment:
Isn't `_.isInstanceOf[JsBoolean]` identical to `_: JsBoolean`? I would have thought they both compile to exactly the same byte code - `isInstanceOf` is a synthetic method that scalac will compile to a `instanceof` byte code instruction. Likewise, `_: JsBoolean` will also compile to an `instanceof` byte code instruction. Both are null safe. The new change is possibly slower because it will involve some jmp statements etc, rather than just directly returning the result of the `instanceof` instruction.